src/stream/src/executor/backfill/no_shuffle_cdc_backfill.rs

```diff
@@ -0,0 +1,420 @@
+        upstream_table: ExternalStorageTable,
+        upstream: BoxedExecutor,
+        output_indices: Vec<usize>,
+        _progress: Option<CreateMviewProgress>,
+        schema: Schema,
+        pk_indices: PkIndices,
+        metrics: Arc<StreamingMetrics>,
+        chunk_size: usize,
+    ) -> Self {
+        Self {
+            info: ExecutorInfo {
+                schema,
+                pk_indices,
+                identity: "CdcBackfillExecutor".to_owned(),
+            },
+            upstream_table,
+            upstream,
+            output_indices,
+            actor_id: 0,
+            metrics,
+            chunk_size,
+            actor_ctx,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(self) {
+        // The primary key columns, in the output columns of the upstream_table scan.
+        let pk_in_output_indices = self.upstream_table.pk_in_output_indices().unwrap();
+
+        let pk_order = self.upstream_table.pk_order_types().to_vec();
+
+        let upstream_table_id = self.upstream_table.table_id().table_id;
+        let upstream_table_reader = UpstreamTableReader::new(self.upstream_table);
+
+        let mut upstream = self.upstream.execute().peekable();
+
+        // Current position of the upstream_table storage primary key.
+        // `None` means it starts from the beginning.
+        let mut current_pk_pos: Option<OwnedRow>;
+
+        // Poll the upstream to get the first barrier.
+        let first_barrier = expect_first_barrier(&mut upstream).await?;
+        let init_epoch = first_barrier.epoch.prev;
+
+        // Check whether this parallelism has been assigned splits,
+        // if not, we should bypass the backfill directly.
+        let mut invalid_backfill = false;
+        if let Some(mutation) = first_barrier.mutation.as_ref() {
+            match mutation.as_ref() {
+                Mutation::Add { splits, .. }
+                | Mutation::Update {
+                    actor_splits: splits,
+                    ..
+                } => {
+                    invalid_backfill = match splits.get(&self.actor_ctx.id) {
+                        None => true,
+                        Some(splits) => splits.is_empty(),
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        if invalid_backfill {
+            // The first barrier message should be propagated.
+            yield Message::Barrier(first_barrier);
+            #[for_await]
+            for msg in upstream {
+                if let Some(msg) = mapping_message(msg?, &self.output_indices) {
+                    yield msg;
+                }
+            }
+            // exit the executor
+            return Ok(());
+        }
+
+        tracing::debug!("start cdc backfill: actor {:?}", self.actor_ctx.id);
+
+        // TODO(siyuan): restore backfill offset from persistent state
+        let (backfill_offset, is_finished) = (None, false);
```

Comment:
Hmm, this sounds critical. So, every time the cluster recovered, it will do backfilling again from the beginning?

Reply:
Thanks for reminding me. I fixed this by saving a backfill finish flag into the state table of source. https://github.com/risingwavelabs/risingwave/pull/11707/commits/4e1ad92e578df5991fe6e0a640b4fec682c9f782
```
dev=> select * from __internal_orders_rw_4004_source_5006;
 partition_id  |                                                                                                                                                                       >
---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------->
 5005          | {"split_info": {"mysql_split": {"inner": {"snapshot_done": true, "split_id": 5005, "start_offset": "{\"sourcePartition\":{\"server\":\"RW_CDC_5005\"},\"sourceOffset\">
 5005_backfill | true
 ```